semver/semver.ts

```diff
@@ -0,0 +1,61 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * A SemVer object parsed into its consituent parts.
+ */
+export interface SemVer {
+  major: number;
+  minor: number;
+  patch: number;
+  prerelease: ReadonlyArray<string | number>;
+  build: ReadonlyArray<string>;
+}
```

Comment:
I think they should be `readonly` there is not a good reason they were not marked as readonly except that the type evolved from the original class. I'll add the `readonly` keyword.

Reply:
For someone who wants to mutate versions for some reason I think there would be a couple of reasonable options, beyond just using the increment api in this library already:


```ts
const v0 = { major: 0, minor: 1, patch: 0, prerelease: [], build: [] };
v0.major = 5; // its not readonly here
console.log(semver.format(v0));
```

```ts
const v0 = semver.parse("0.1.0");
console.log(semver.format({ ...v0, major: 5 ));
```

Reply:
> For someone who wants to mutate versions for some reason I think there would be a couple of reasonable options, beyond just using the increment api in this library already:

@justinmchase That's true when the value is not an immutable object returned by this library, but when working with a type that conforms to the fully readonly `SemVer`, it's more complicated — and that's why I suggested providing a helper function for getting a mutable copy. Here's an example to illustrate:

```ts
type SemVer = {
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
  readonly prerelease: ReadonlyArray<string | number>;
  readonly build: ReadonlyArray<string>;
};

declare function parse(version: string): SemVer;

// Get an immutable SemVer from a function in this module:
const semver = parse("v1.1.1");

// Can't mutate:
semver.minor = 2; /* Error
       ~~~~~
Cannot assign to 'minor' because it is a read-only property.(2540) */

// Types used by function below:
type Mutable<T> = { -readonly [K in keyof T]: T[K] };
type MutableSemVer = Mutable<{ [K in keyof SemVer]: Mutable<SemVer[K]> }>;

// Help a user get a mutable copy of a SemVer
// The fn name is just a Rust naming convention (not necessarily a suggestion)
function toMutable(semver: SemVer): MutableSemVer {
  return structuredClone(semver);
}

{ // Easy for user with a helper function and type: 🤩
  const semverMut = toMutable(semver);
  semverMut.minor = 2; // Ok
}

// Complicated for user to do correctly without those: 🥵

{ // Either tedious nested spread syntax (and prone to
  // breakage if the SemVer structure should ever change):
  const semverMut = {
    ...semver,
    prerelease: [...semver.prerelease],
    build: [...semver.build],
  };
  semverMut.minor = 2; // Ok
}

{ // Or complex type manipulation to annotate a correct, narrow return type
  // because the return type of the structuredClone function is `any`:
  const semverMut = structuredClone(semver) as {
    -readonly [K in keyof SemVer]: SemVer[K] extends readonly any[]
      ? SemVer[K][number][]
      : SemVer[K];
  };
  semverMut.minor = 2; // Ok
  //^? const semverMut: { major: number; minor: number; patch: number; prerelease: (string | number)[]; build: string[]; }

  const defaultValue = structuredClone(semver);
      //^? const defaultValue: any
}

```
[playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&target=99&jsx=4&exactOptionalPropertyTypes=true&isolatedModules=true&noImplicitOverride=true&ssl=52&ssc=5&pln=1&pc=1#code/C4TwDgpgBAyhC2A1CAnKBeKBvAUFKKEAhgCYD2AdgDYhTxEBWZKAXFBQK7wBGqA3HgLFy1WvACWFZm049+gwqUo0oYIsADGACxldeKAfkUiVYQoSrEAzhDYAlYcpABBFCiIgAPFeApJAcygAH3Y9VAA+QyElUShuDnEqEntHUVd3Lx8-Cn9InABfARwSCA0qIkIoADMOCg1gcUpVCpsACgA3VCtGijYsgIBKNjgkeRwAenGoAHEIYCgiCihxeHgOYCJuS1gEZDQqlDJ4Bera+p7lpeAtcSs6MhIOSxYcDUofKBt4TrRMNRQ2gAidoARgAdOCQYCBkVJlAAMKLADk8zWG2AthwXx+YIkUl+UAATHwoOMAFRQACibmYgnw9IAfkymThERQpPMiFZuv4rmQoEi8cwkXFSkQODZlvNbidjABaJyqQ6QFCgMGtQkAVgALAAGAZQMnjHATKYAFXAEDuEogJDitBqdQaTV4VDIAHcXqBIFAALLrTaWTxm8IYbBQOXGRUAbQA0pcoABrCAgMhVKBmgC6bDNcczUEKOG90H9Gy2EBGezDpcDEE8WCgcYTydT6crqGzfoD5c87ZQedD+TypqgAAkIFQwCcbWh-HMTmja1A3mBaGmTn2R2atNAqksKER4NAZQwJZyoHYz+xDwFl5ROhRnUtWhz2KUrVYKolaERPhx-HOPg9AMOCOucTTAGQNblq02KoMMuyoEMXZlpYfbYAocwcCgSxZBw9TYba8JuhQECwQgPwwgUJoNnClJcg6zBQDOUDuuI1wnDuk6oKcToXIsdrFmwgB8G4AlLuCG8FAfHBKClmGkHQZY5HfEhUQyaWuKSExmDEqSUwAPKJtRI7wkcYBUOIGjqLa1RMSxkFQOQd5uKUwAqGx1xkOsUCeTYImAK67NF6VS7E7mgGIkI0Epvj4NlWGYwifCAj5EAAHlArQCUqlDQJBghwtwiiJkQc7LOm1zQOheEEZUVhaF5SRQBAPzLloixzkMEnvPM6neZguD0lAYLDTJAA0dJKqgE7WLYjbDWCMlgglFgzZm42DfEiTJHNI0UagYKbUka2CIW+C9cAmn4mGulwoZxm0QZaBvPA5kQOlxZ0Is4hgE86gXA5iwctZJxvC59SjdeNLukIwDYVclr5VMvBWTaPk7jDcM+ZaUDrhVny+PhsOECQxE5bx4FLDKAAGiwgFTLz4JJ0l7bJfX4yghOESTJFkTJBpchhg2RqkKhNpISYpuufadn2eZNalGIUCQdxRrEtPRpmE34AA-Dsoz9rGmbRrI+hG5rg34Ah+t5lEp2fCzGlCgSN0GUZ+CTAAetrJqM91jkQFU4pUMAiBEFQHDQJg1VE0RPPKZRUSDZ73v5DgQA)

Reply:
Yeah I think you're right that its kind of annoying that its immutable, and frankly none of these functions actually care about mutability anymore since they're all stateless. So I think I'll just remove the `readonly` keywords, none of these functions will mutate your objects but if you want to go right ahead.